Updated environ.sh.sample CFLAGs.

gyrovorbis opened the pull request:
1) Link-Time Optimization
    - Included -ffat-lto-objects by default, which is required for atomics and many KOS ports, with a note about this. 2) Additional Optimizations
    - No longer is about LTO
    - Has a list of random flags which have been found to also positively benefit perf at least SOMETIMES but are potentially situational.
3) Position Independent Code
    - Disabled by default with explanation of when to enable (dynamic linking) 4) Frame Pointers
    - Added info on using an extra register for frame pointer 5) GCC Builtin Functions
    - enabled by default now for performance reasons, as it has not been found to cause any issues (and our compiler doesn't do much with it)
    - Modified warning to be less harsh, as at this point anything broken with this flag is a new bug, most likely. 6) Fast Math Instructions
    - Added comment advising people to still try to enable these flags on individual compilation units rather than globally, sure enabling globally cause any issues.

Commit:
Don't block for rumble and vmu beep/screen commands

There simply isn't anyone that is relying on this
blocking behavior and all users end up working
around it. The nature of the maple queue makes it
so that the blocking time is significant and there
really isn't much of anything that an end user could
do in the case of a failure.

Though I had originally considered (in #786) adding
new functions to select blocking/non-blocking there
doesn't seem to be any reason for it. If some use
cases are discovered in the future, then the API
expansion should be to add intentionally blocking
versions.

Commit:
Remove unneeded defines.

Commit:
Whitespace and comment cleanups.

Somehow there was quite a bit of stray whitespace.

Commit:
Catch undefined exceptions as panic.

This should not happen.

pcercuei committed:
dmac: Drop old DMAC interface and header

And move the code initializing DMAC on Naomi to the DMAC driver.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

darcagn committed:
Eliminate non-standard fixed-width integer types from examples

Commit:
Update Changelog and Relnotes for v2.2.0

gyrovorbis committed:
Added CMake installation include and lib dirs.

Modified dreamcast.toolchain.cmake to automatically configure CMake's
two environment variables controlling where files get copied to when
installing (CMAKE_INSTALL_INCLUDEDIR, CMAKE_INSTALL_LIBDIR) to
automatically point to the proper locations within kos-ports.

gyrovorbis committed:
Addressed review feedback (kinda).

gyrovorbis committed:
CMAKE_INSTALL_BINDIR => DC_TOOLS_BASE.

darcagn committed:
Use THD_STACK_ALIGNMENT to adjust arch's stack alignment requirements

Commit:
Fix various warnings in examples.

Uninitialized variables, unused args, and signedness
mismatches.nehe26 got a bit deeper of a cleanup in
order to address unchecked allocations.

Commit:
stack: Don't deref fp twice.

In #600 there was a change to store `arch_fptr_ret_addr` as
it was being tested as valid before being printed. Unfortunately
this was done in a way that overwrote the current fp being worked
on. So when it came time to pass it to `arch_fptr_next` another
layer of deference was happening.

Commit:
irq: Don't deref fp twice.

In #600 there was a change to store `arch_fptr_ret_addr` as
it was being tested as valid before being printed. Unfortunately
this was done in a way that overwrote the current fp being worked
on. So when it came time to pass it to `arch_fptr_next` another
layer of deference was happening.

Commit:
arch: Mark register access inlines as `always_inline`

The behavior of these would break if the function call
is not inlined, so ensure that they are always inlined.

Commit:
Apply noinline if we expect a function call for stack tracing.

In the examples, ensure that the functions we are stack tracing
are not inlined as it would defeat the purpose of the example.

Inside the stack tracing functions and assert, there are assumptions
made that we are going back by fixed numbers of frame pointers, so
avoid inlining as that would the intended tracing.

pcercuei committed:
dc-chain: Support multiple archs and decouple ARM

Instead of having all the scripts handle two architectures
(SH4 and ARM), with all rules duplicated, update them so that they will
support the architecture specified in the "platform" variable.

This means that the exact same scripts can be used for building the SH4
toolchain, and also the ARM toolchain. Later, it will also allow
building toolchains for other platforms.

One interesting side effect of this change is that a C library is now
built for the ARM.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
utils: Add kos-gdb wrapper

This is a wrapper around gdb-multiarch, which sets a few settings so
that it can be used for Dreamcast development.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Add support for PPC toolchains and add GameCube profile

Add support for building a powerpc-eabi toolchain, and add a profile for
the Nintendo GameCube.

Update the KOS patch for newlib and cherry-pick two patches from master
to fix building on PowerPC.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
environ.sh: Export $KOS_GDB_CPU variable

This variable is defined to the CPU architecture name used with
gdb-multiarch.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Move profiles to platform subdirectory

To ease the introduction of multiarch in dc-chain, move the current
profiles to a dreamcast/ subdirectory.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Make per-platform Makefile.cfg

Rename Makefile.default.cfg to Makefile.dreamcast.cfg.

Later, when we support more platforms, we can add
Makefile.<platform>.cfg.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Add AICA toolchain profile

Add a toolchain profile to build a toolchain for the Dreamcast's AICA
sound co-processor.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

Commit:
Update gitignore to ignore generated `authors.h`

DC-SWAT committed:
cdrom: Fixed confusion between sessions and areas.
GD-ROM drive does not differentiate sessions at all. This TOC is always divided into two, for the low-density area and the high-density area, which is only on the GD-ROM disc.
And in fact, this is one and the same TOC, it’s just that for non-requested area the tracks are erased (and first/last changed) in this common TOC, because the tracks numbering remains the same.
There may be multiple sessions in one area, but there is no division between them.

ppxxcc committed:
Update to detect Apple Clang 17+ and fix zlib issue.

New Apple Clang 17 in Xcode 16.3+ causes build issues with bundled in zlib in binutils/gcc, so detect compiler version and use system zlib if necessary.

Commit:
Add defines for various screen layout constants

pcercuei committed:
cdrom: Add function cdrom_poll(), use it in cdrom_req_cmd()

The cdrom_poll() function will be used every time we need to poll the
GD-ROM drive using syscalls.

Use it in cdrom_req_cmd(), which changes from a 10-turn poll to a poll
with a 10ms timeout, which should be plenty of time.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
cdrom: Use cdrom_poll() in cdrom_exec_cmd_timed()

Factorize code by using cdrom_poll() in cdrom_exec_cmd_timed().

This commit also fixes cdrom_exec_cmd_timed() being extremely slow;
that's because after the command was sent to the GD drive, the response
was only obtained after the next VBLANK, which meant having to wait for
up to 16ms.

Now, instead of waiting on the "cmd_done" semaphore signaled in the
vblank interrupt, we poll for the command response until we obtain it.
It will cause a slight CPU overhead but will be much faster.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
cdrom: Use cdrom_poll() in cdrom_abort_cmd()

Factorize code by using cdrom_poll() in cdrom_abort_cmd().

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
cdrom: Use cdrom_poll() in cdrom_get_status()

Factorize code by using cdrom_poll() in cdrom_get_status().

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
cdrom: Use cdrom_poll() in cdrom_read_sectors_dma_irq()

Factorize code by using cdrom_poll() in cdrom_read_sectors_dma_irq().

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
cdrom: Use cdrom_poll() in cdrom_stream_stop()

Factorize code by using cdrom_poll() in cdrom_stream_stop().

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
cdrom: Use cdrom_poll() in cdrom_stream_request()

Factorize code by using cdrom_poll() in cdrom_stream_request().

Note how the calls to syscall_gdrom_dma_check() and
syscall_gdrom_pio_check() are replaced by a call to
cdrom_stream_progress() for further factorization.

Also, the ERR_NO_ACTIVE value returned by the poll callback function is
just here to signal the code calling cdrom_poll() that we got a
"NO_ACTIVE" or "COMPLETED" response.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
cdrom: Simplify vblank code

Since nothing waited on the "cmd_done" semaphore anymore, the code
signaling it as well as the semaphore itself can be dropped.

The timeout code can also be dropped, as it was now handled by the
cdrom_poll() function.

The VBLANK interrupt handler now only has to check for errors when DMA
is used.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

gyrovorbis committed:
Fixed maple crash when unknown device attached.

Currently, the Maple driver crashes due to NULL pointer dereference in
"maple_driver.c" on line 121. When the user manually disables certain
Maple drivers via explicitly providing a custom KOS_INIT_FLAGS() list,
and such a device is plugged into the controller, the "dev" pointer
never gets allocated, since a driver for it is never found.

1) Check to see if dev is NULL and error out early before dereferencing
   it.

gyrovorbis committed:
Updated environ.sh.sample CFLAGs.

1) Link-Time Optimization
    - Included -ffat-lto-objects by default, which is required for
      atomics and many KOS ports, with a note about this.
2) Additional Optimizations
    - No longer is about LTO
    - Has a list of random flags which have been found to also
      positively benefit perf at least SOMETIMES but are potentially
situational.
3) Position Independent Code
    - Disabled by default with explanation of when to enable (dynamic
      linking)
4) Frame Pointers
    - Added info on using an extra register for frame pointer
5) GCC Builtin Functions
    - enabled by default now for performance reasons, as it has not been
      found to cause any issues (and our compiler doesn't do much with
it)
    - Modified warning to be less harsh, as at this point anything
      broken with this flag is a new bug, most likely.
6) Fast Math Instructions
    - Added comment advising people to still try to enable these flags
      on individual compilation units rather than globally, sure
enabling globally cause any issues.

pcercuei commented:
@gyrovorbis NAK on the `-fno-PIC -fno-PIE`. It is a good idea, just not a day before 2.2.0. So I'd suggest to fork it to a new PR that can be merged after 2.2.0 is out.

darcagn commented:
I find the phrasing around LTO confusing. What would the user think "fatter object files" implies? It's also reasonable to conclude when reading "atomics REQUIRE fat LTO objects to work" that atomics require LTO.

I would leave in a warning about LTO. It doesn't have to be the exact phrasing `Certain KOS-ports and examples do not work properly with "-flto=auto"!` but it's well known to cause build issues and uncover subtle bugs on any platform, not just Dreamcast (Ubuntu, for example, maintains a list of packages that are not to be compiled with LTO), so users should be aware of reasons not to enable it (and if it were the case that there are none, then we'd be enabling it by default). It also increases compile times and memory utilization although that is much less of a concern when targeting Dreamcast.

The "Additional Optimizations" section should have each flag on its own line and not refer to it as a "set." The optimizations are independent of one another and should be easily toggled on and off individually, unlike most other options in the environ where the flags go together.

Commit:
Minor type cleanup and restructuring of strangely organized logic.

Commit:
Give a message if no controller is detected, and don't write if no vmu found.

Commit:
Create define for data size and ensure the pkg is filled out.

Commit:
vmu_pkg: add bounds checks.

Avoid memcpy from NULL and accepting a pkg with an invalid
`data_len`.

Commit:
Map keyboard 'return' key to CR not LF.

This was a regression in the new API from the start in 2012.
By mapping LF to 'return' (0x28, the 'enter' key in the main
block of keys) it meant that there was no way to send a return
on keyboards that lacked a numpad.

This change rolls us back to the mapping that had always been
used in the older API. It seems that the issue had not been
encountered or noted in large part because the old API's mapping
had been maintained in code until #984 where I removed them for
only differing in this one character code from the standard US
keyboard. The old API is still what was backing conio/kosh and
the change made them difficult to use.

Of course it's still possible for the user to distinguish between
the keys if they'd like, and these mappings only apply to using
our built-in ASCII translation.

Commit:
Ensure that the thread's return can fit in the variable provided.

`thd_join` is meant to take a `void **` by passing instead `bool *`
you could get all sorts of bad stuff when the data overflows the bool.

Commit:
Update version to 2.2.0.

Additionally clean up some whitespace.

Commit:
Keep network examples connected to dcload console.

By default most network examples were set to call dctool
with `-n` to disconnect the debug console. The idea behind
it seems to have been to prevent dcload-ip from interferring
with the network traffic, but that's not needed. We have a
system in place just for this and it relies on the dcload
console being active in order to request the already assigned
IP.

Commit:
Use `__is_aligned` and `__alignup` builtins.

To note, I've applied them only when the param is a pointer.

Commit:
Use regfield macros.

This actually makes a difference in asic as there was a possibility
of shifting up by 32, enough to be invalid for a signed int and the
BIT macro enforces `1u`.

pcercuei committed:
dc-chain: Move gthr-kos.h to non-arch-specific path

The gthr-kos.h is not arch-specific. To avoid having to hardcode
different paths for different archs in the GCC patches, install the
gthr-kos.h file to an arch-independent path in libgcc.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

Commit:
Clean up usage of packed attribute in addons.

Both KOS and newlib provide `__packed` as a compatibility
layer, no reason the addons should be operating differently.

Commit:
arch/args.h: Add standard double-include protection and decls

darcagn committed:
dc-chain: Make GDB CPATH and LIBRARY_PATH exports for macOS conditional
In the default case, we specify the CPATH and LIBRARY_PATH
environment variables for users using the Homebrew package
manager, which is the most common case. By making these
variables conditional, we allow users of other package
managers to specify their own environment variables here.

Commit:
Fixes for doxygen warnings

pthread.h - endcond was inside the header ifdef guards and cond was
 on the outside, so doxygen couldn't see them both.

vmu.h - wrong ref name was being used for bfont icons.

keyboard.h - There may be a better fix for this, but it seems that
 doxygen is choking on the depr inside a union.

pcercuei committed:
dc-chain: Rework patch folders and support target-specific patches

Differenciate between the target-specific and host-specific patches to
keep the patches folder clean.

Add a target-specific search path for patches, so that it is possible to
conditionally patch the code according to the target arch.

Drop the execution rights on the newlib-4.3.0.20230120-kos.diff patch.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Replace arch-specific fake-kos.S with C code

The fake-kos.S files were only defining weak symbols as aliases to a
dummy function. This can be done in C as well; the advantage being that
the same C code can be used on all architectures.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Drop old KOS defines, move the rest to common GCC config file

The toolchain will only now just provide the __KOS_GCC_PATCHLEVEL__
macro. From the version specified there, KallistiOS can then figure out
whether or not it supports 32 MiB of RAM.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Add patches for GCC 8.5.0

GCC 8.5.0 is the last one that supports the ARM7DI processor.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

Commit:
examples: replace `.keepme` with `.gitignore`

In each case that we were using `.keepme`, we had an empty romdisk
folder that is either populated by generated files, or in the case
of the sound examples user provided files. Instead of having a
`.gitignore` in the main folder of the example that lists the files
to ignore in the romdisk, then a separate `.keepme` just to keep the
dir alive, we can have the `.gitignore` be directly in the romdisk
folder. To match this, the Makefile.rules was updated to have romdisk
building ignore `.gitignore` (which seems like just a nice thing
anyways for users).

Additionally, the sound examples previously did not have any `.gitignore`
so so it was added to prevent the `test.mp3/ogg/opus/adx` files from
appearing as untracked files.

pcercuei committed:
dreamcast: Add support for toolchains without __KOS_GCC_32MB__

Newer toolchains don't define __KOS_GCC_32MB__ anymore as KOS does not
support any of the older toolchains not capable of 32 MiB RAM support.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
Fix built object file not added to the KOS library

The source files inside the kernel/ subdirectory (which is just
version.c for now) were compiled, but the object files were not added to
the KallistiOS library.

Address this by adding a Makefile dependency on the "defaultall" rule,
which will copy the object files to the build folder, so that they are
subsequently compiled inside the KallistiOS library.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

darcagn committed:
dc-chain: Change source URL for config.guess and config.sub to more reliable mirror

pcercuei committed:
environ.sh: Drop $KOS_START

This was only used with GCC < 4.0, which we don't support anymore.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
Rework how startup code is handled

Instead of building the arch's startup.o and copying it inside the
toolchain as crt1.o, point the linker script to the correct startup
code from the KallistiOS lib folder.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Don't override the toolchain's crt1.S

This is not needed anymore, as the linker script will pick the startup
code itself.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Do shallow clones of git repositories

We don't need the whole history tree of the git repositories we want to
clone. Shave off the download size (and therefore the build size) by
performing shallow clones.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Detect and use number of CPU cores

Instead of hardcoding the default number of jobs to 2 (which is very low
by today's standards), leave the 'makejobs=' variable empty by default,
and update the Makefile scripts to automatically detect the number of
logical CPU cores, and use that as the number of jobs.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
examples: Add fb_tex example

This example demonstrates a trick that allows the framebuffer to be used
as a texture.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
video: Add vid_set_dithering()

This function can be used to enable or disable dithering.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
pvr: Limit texture addresses to VRAM size

Previously, the various compile functions would apply a 0x00fffff8 mask
to the texture address. This is strange as it limits the address space
to 16 MiB while the Dreamcast has only 8 MiB of video RAM.

Use a new mask of (PVR_RAM_SIZE - 1) to address this issue, and make it
more obvious what the mask is for.

Note that the low 3 bits don't need to be cleared as the address is
shifted right by 3 later.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
pvr: Update pvr_get_front_buffer()

Previously, the VRAM offset of the front buffer was doubled, and then
masked to the VRAM size. This effectively meant that for a front buffer
at VRAM address 0x400000, we would get VRAM address 0x0. This actually
works, as the front buffer lies in 32-bit space while textures lie in
64-bit space; the framebuffer at that address could then be used as a
texture from VRAM address 0x0 with specifically crafted U/V values.

The problem with the previous implementation is that due to the masking,
the calling code has no way to know if the front buffer resides in the
lower 4 MiB of VRAM, or the upper 4 MiB of VRAM - and this information
is needed for using the front buffer as a texture.

Update the function so that it does clamp the front buffer address to
the VRAM size before the doubling, and not after. With the example
above, the address computed for the front buffer would be 0x800000,
which is bigger than the VRAM size. This does not matter, as the poly
compile functions will clamp the texture address to the VRAM size. But
it will allow the calling code to know if the front buffer is in the low
or high VRAM.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Add 15.0.0-lra toolchain profile

This toolchain profile can be used to test a toolchain with GCC's new
LRA register allocator instead of the old one.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
pvr: Extend VRAM to 16 MiB on Naomi

The Naomi has 16 MiB of VRAM.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
timer: Fine-tune timer rate

The seconds counter has been reported to drift quite significantly vs. a
wall clock. This is most likely because until now we considered that the
peripheral clock was a perfect 50.0 MHz, while in practice it is closer
to 49.9 MHz.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

DC-SWAT committed:
isofs: Added cdrom streams and direct DMA support (#763)

* Added cdrom DMA streams support. At the sector boundary with aligned buffer, DMA streaming will be started for the entire file and works until file end or cancelled (closed or another file used). Reading from an open stream can be done in chunks of up to a minimum of 32 bytes.
* Added IOCTL function to determine that file system can load directly to SPU and PVR RAM's thought the Root Bus.

pcercuei committed:
dc-chain: Remove multiple files in one single command

There's no need to run multiple "rm -rf" commands when a single one can
do the job.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: doc: Update changelog

Add a description of my changes in the changelog.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: doc: Update path of default Dreamcast Makefile config

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

darcagn committed:
Fix net_init crash when there are 0 detected network devices

Commit:
Update version number for current master.

Moving forward I believe this should be the way we
operate. After a version is cut, we set master to
the next version patch version. This should allow
for more consistent ability to test with the
presumption that master is a higher version than
the most recent one.

Commit:
vmulcd: Correct data type of vmu fb.

There was, for some reason, a mix of the signed `char` and
unsigned `uint8_t` as the types for the vmu fb. This caused
possible unsigned conversion warning/errors when manipulating
the data.

Switched to `uint8_t` universally as the usage is for data we
need to have specified by the width of it.

Commit:
net: Replace `arch/types.h` types with stdint.

Commit:
dmac: Avoid bug with `__is_aligned` macro.

`__is_aligned` as defined in newlib was missing parens around
the first parameter, so adding them here at the call to
correct its usage. This has been fixed upstream but requires
a toolchain update to be effective.

gyrovorbis committed:
Addressed review feedback.

pcercuei committed:
CMake: Use absolute path for KOS variables

The CMAKE_INSTALL_BINDIR and CMAKE_INSTALL_LIBDIR require absolute
paths. Address this issue by using absolute paths for all KOS variables
that contain paths.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
pvr: Limit texture address to VRAM size in to_pvr_txr_ptr()

The previous commit f2513c1e1 ("pvr: Limit texture address to VRAM
size") missed one instance, the one in to_pvr_txr_ptr().

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
pvr: Use to_pvr_txr_ptr() to factorize code

Use to_pvr_txr_ptr() where it makes sense to factorize code.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Add __is_aligned() macro fix to Newlib patches

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Add missing symbols to fake-kos.c

Add missing symbols that are needed for libobjc to compile properly.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
dc-chain: Drop useless change in GCC patches

Those are not needed anymore as the fake-kos.c file provides the symbols
needed to compile and link a program that calls printf().

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

GPF committed:
Fix keyboard LED state handling and add keyrawtest example (#1095)

* Fix keyboard LED state handling and add keyrawtest example

This change improves keyboard driver behavior by correctly tracking
Caps Lock, Num Lock, and Scroll Lock LED states. It also adds a
keyrawtest example which captures raw keyboard events using
kbd_queue_pop(dev, 0), logs printable and special key inputs, and
displays LED state changes in real time. Useful for debugging
non-ASCII keys and verifying modifier behavior.

GPF committed:
CMake: Set default installation directories only if not defined

Commit:
dmac: Restrict dma_init back to NAOMI.

Reverts the behavior change from #983 that had
KOS performing the same initialization steps on
DC as NAOMI. This was breaking PVR dma as seen
via the dma speedtest example.

GPF committed:
Add missing include path for architecture-specific addons

Commit:
pvr: Correct bits for strided texture format.

This has been incorrect since added. Bit 21 of the
texture format apparently has no effect, and bit 25
is what sets whether or not the texture u size is
read as a power of 2 or 32x the value in the lower bits
of the PVR_TEXTURE_MODULO register.

This striding does not apply for YUV textures and so
the use was simply removed in those examples.

Commit:
pvr: Add getter/setter for strided texture value.

Provide a friendly API to get/set the currently configured
texture stride value. That value is used to set the stride
width for any texture using the `PVR_TXRFMT_X32_STRIDE` flag.

These were originally authored by Andress Barajas for #836.

Additionally update the `fb_tex` example, to use this API
rather than directly modifying the register.

Commit:
pvr: Add example for using 32-pixel strided textures

Originally authored by Andress Barajas for #836.
Demonstrates the generation of a 640 width texture
and the rendering of it using the global strided textures
setting.

Commit:
pvr: Replace arch/types usage in pvr_texture.

Additionally replace an alignment rounding with
the `__align_up` macro.

Commit:
fs_romdisk: Replace arch/types with stdint.

Additionally correct minor whitespace issues around
var declarations or casts.

Commit:
fs_romdisk: Update param to `bool`.

Commit:
fs_romdisk: Remove header pointer from internal image struct.

This is always simply the start of the image, and on top of
that it is never used. So just removing it. Cleaned up the
access to the magic string from the header as well.

Commit:
fs_romdisk: Refactor `fs_romdisk_unmount`

Use the newer scoped mutex lock, reduce nesting, deduplicate
the own_buffer check, and remove unneded `rv`.

Commit:
fs_romdisk: Refactor romdisk unmounting/shutdown.

`fs_romdisk_shutdown` was effectively duplicating
`fs_romdisk_unmount` in a loop over all entries in
the `romdisks` list. Should be functionally identical
except for now `fs_romdisk_shutdown` both locks the
list mutex while working on it and removes the entries
from the list.

Commit:
fs_romdisk: Tidy romdisk_find.

Was strangely using a nested else{if/else} rather than elseif/else.

Commit:
fs_romdisk: Deduplicate fd validity checks.

For all but the two checks in open/close, this is
just an identical replica of what was happening.

In open, if the fd is in bounds, then index must have
already be set to `FH_INDEX_RESERVED` so it can't
make a difference. In close, if it's invalid because
it's already set to `FH_INDEX_FREE` then there's nothing
to be done anyways.

Commit:
fs_romdisk: Dynamically allocated FD tracking.

As with #973 remove the static table and pass a pointer
to the FD object out rather than index.

This additionally means that the define for the max files
can be removed. While there I removed the one for CDFS
that should have been removed when it was obsoleted.

Commit:
fs_romdisk: Free dangling FDs on shutdown.

It's just polite.

pcercuei committed:
thread: Simplify thd_schedule()

Remove the 'now' parameter which was never really used.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

pcercuei committed:
thread: Factorize genwait wake code

Factorize the genwait wake code instead of having two slightly different
functions that do more or less the same thing.

Use TAILQ_FOREACH_SAFE() to simplify the code further.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

Commit:
fs_elf: Move to stdint types.

Commit:
elf: Typedef structs and clean up whitespace.

Commit:
elf: Replace custom debug output with dbglog.

I've set it to `KDEBUG` so it's only if you want
all the output.

Commit:
elf: Check for `fs_read` failure.

Commit:
elf: Add and use header ident defines.

Additionally clean up the ident code slightly by
providing fixed character width to the printf
rather than doctoring the header to NUL terminate.

Commit:
elf: Refactor and update header validation.

Move header validation to its own internal function.
Additionally remove magic numbers by adding defines
to back the values in the header.

Finally, expanded the number of internal defines being used
per-arch so that it includes the bit width and endianness
fields. This should make it simple to add new ones, though
I removed the old archs at the same time.

Commit:
elf: Expand Sections and section flag defines

These are all aspirational. Also fixed a typo.

Commit:
Remove deprecated header.

Was replaced in 2017 by implementing `<sys/uio.h>`.

Commit:
dbglog: Lower the default debug level.

`DBG_INFO` is noted as being a good level, so lets use that.
It never made much sense to have this default to the highest
setting, and it's a frequent user complaint.

Commit:
dbglog: Add option to compile-time disable/limit debug logging.

This should help allow debug messages to be garbage collected,
and provide simpler steps for a release build when paired with
NDEBUG. `DBGLOG_LEVEL_SUPPORT` can be set to a dbglog level in
order to cap the level available at runtime to that amount.
`DBGLOG_DISABLED` overrides this down to disallow any output.

As the actual function backing `dbglog` is now named differently
this had to be updated in the exports. I've additionally moved it
out of the stdio section to reflect that it's kos-internal and
no longer is being shared with stdio after #1000 .

Commit:
dbglog: Provide functionality for source-specific verbose debugging.

We provide many of these throughout the codebase where a single
driver, subsystem, or file will have its own define to set if
extra verbose or detailed debug info is outputted for it. These
have been cumbersome to use as they rarely had a unified style
or mechanisms, and presumed the defined would be set in the
individual source file.

Even once these options were made available in opts.h, it was
quite inconsistent as each implementation might use different
mechanisms (printf vs dbglog, for instance).

Now, instead, dbglog allows a 'level' to be passed which takes
a define that if defined means output regardless of the set level,
and if not defined will only output in the new 'DBG_MAX' setting
which explicitly has all dbglogs output.

Commit:
fs_vmu: Use new dbglog source functionality for verbose debug.

Additionally add the include for <kos/opts.h> which was missing.
This was getting chained in pretty much accidentally from kos/fs.h.

gyrovorbis commented:
What do you guys think now?

pcercuei commented:
@QuzarDC Using `sinf()` and `cosf()` with builtins enabled will resolve to `__builtin_sinf()` and `__builtin_cosf()`; without `-ffast-math` they won't use the `FSCA` opcode, but resolve to a call to a C library function.

With builtins disabled, they will directly resolve to calling the same C library functions.

Therefore disabling builtins by default without a valid reason is a bad idea and will make the code *potentially* slower, and not faster.

pcercuei committed:
dc-chain: Update build instructions

Update the build instructions in the README.md to reflect the new
changes.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

darcagn committed:
Move arch-specific ELF definitions from elf.c to arch.h

gyrovorbis committed:
Merge branch 'master' into falco/environ_sample_update

gyrovorbis committed:
Fixed "library" example to build with LTO enabled.

...by disabling LTO for them explicitly, which is fine when using
-ffat-lot-objecst.

darcagn commented on doc/environ.sh.sample:
```diff
@@ -172,21 +191,21 @@ export KOS_CFLAGS="${KOS_CFLAGS} -fomit-frame-pointer"
-# compiler-optimized implementations to replace standard function invocations.
-# The downside of this is that it COULD interfere with Newlib or KOS implementations
-# of these functions, and it has not been tested thoroughly to ensure compatibility.
+# Comment in this line to prevent GCC from using its own builtin implementations of
```
```suggestion
# Uncomment this line to prevent GCC from using its own builtin implementations of
```

gyrovorbis committed:
Update doc/environ.sh.sample

Co-authored-by: darcagn <darcagn@protonmail.com>

gyrovorbis committed:
Merge branch 'master' into falco/environ_sample_update

QuzarDC commented:
> Also, on the subject of builtins, I concur with @pcercuei. You definitely want them enabled by default because they can prevent full function invocations to Newlib for things like special memcpy() specializations (which we have almost none of in GCC), or for replacing sinf()/cosf() with FSCA.
> 
> I've only ever built with builtins enabled for many, many months and have never had a problem.

My question was specifically because you raised the issue of not having builtins turned on *unless* `-ffast-math` was also turned on: https://discord.com/channels/488332893324836864/614912396460556298/1333936876200919133 back in January when we tried turning them both on and fast-math (not builtins) caused issues with raylib and something else. At the time we were going to just disable fast-math but you'd raised that the math builtins have worse performance without fast-math than without builtins at all. 

If I misunderstood, or it's changed since then that's fine. If not, would just be a good thing to note that they should/need to be used together. I know we haven't had any compatibility issues in ages (iirc back when it was occasional linking errors and stdio breakage), I just remember there being this other downside.

gyrovorbis commented:
> > Also, on the subject of builtins, I concur with @pcercuei. You definitely want them enabled by default because they can prevent full function invocations to Newlib for things like special memcpy() specializations (which we have almost none of in GCC), or for replacing sinf()/cosf() with FSCA.
> > I've only ever built with builtins enabled for many, many months and have never had a problem.
> 
> My question was specifically because you raised the issue of not having builtins turned on _unless_ `-ffast-math` was also turned on: https://discord.com/channels/488332893324836864/614912396460556298/1333936876200919133 back in January when we tried turning them both on and fast-math (not builtins) caused issues with raylib and something else. At the time we were going to just disable fast-math but you'd raised that the math builtins have worse performance without fast-math than without builtins at all.
> 
> If I misunderstood, or it's changed since then that's fine. If not, would just be a good thing to note that they should/need to be used together. I know we haven't had any compatibility issues in ages (iirc back when it was occasional linking errors and stdio breakage), I just remember there being this other downside.

oh, I see. That conversation was basically me being shocked that `__builtin_sinf()` and `__builtin_cosf()` would not generate the `FSCA` instructions alone without `-ffast-math` and `-mfsca`. It wasn't really that builtins were worse for trig than not using builtins, it was just that using builtins *alone* was not sufficient to emit the `FSCA` fast trig instructions.

That's why the "fast math" line in the `environ.sh.sample` file looks like so:
`#export KOS_CFLAGS="${KOS_CFLAGS} -fbuiltin -ffast-math -ffp-contract=fast"`

Where you can see it also EXPLICITLY enables builtins, because you must combine both builtins *and* fast math to get the desired outcome.

There IS an issue currently with Raylib ICE-ing the compiler on a particular file with `-ffast-math` and builtins enabled, but I have opened a PR to address that already, which hopefully gets merged any day now.

gyrovorbis commented:
Since I fucked up the rebasery, this PR has been replaced by: https://github.com/KallistiOS/KallistiOS/pull/1127
